File: objectModel/TypeScript/__test__/Cdm/Imports.test.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import { CdmDocumentDefinition } from '../../Cdm/CdmDocumentDefinition';
import { CdmManifestDefinition } from '../../Cdm/CdmManifestDefinition';
import { CdmCorpusDefinition, cdmStatusLevel, StorageAdapter } from '../../internal';
import { LocalAdapter } from '../../Storage';
import { testHelper } from '../testHelper';

/**
 * Testing loading imports on a cdm file
 */
// tslint:disable-next-line: max-func-body-length
describe('Cdm/ImportsTest', () => {
    const testsSubpath: string = 'Cdm/Imports';

    /**
     * Does not fail with a missing import
     */
    it('TestEntityWithMissingImport', async () => {
        const localAdapter: LocalAdapter = createStorageAdapterForTest('TestEntityWithMissingImport');
        const cdmCorpus: CdmCorpusDefinition = createTestCorpus(localAdapter);

        const doc: CdmDocumentDefinition = await cdmCorpus.fetchObjectAsync<CdmDocumentDefinition>('local:/missingImport.cdm.json');
        expect(doc)
            .not
            .toBeUndefined();
        expect(doc.imports.length)
            .toBe(1);
        expect(doc.imports.allItems[0].corpusPath)
            .toBe('missing.cdm.json');
        expect((doc.imports.allItems[0]).doc)
            .toBeUndefined();
    });

    /**
     * Does not fail with a missing nested import
     */
    it('TestEntityWithMissingNestedImportsAsync', async () => {
        const localAdapter: LocalAdapter = createStorageAdapterForTest('TestEntityWithMissingNestedImportsAsync');
        const cdmCorpus: CdmCorpusDefinition = createTestCorpus(localAdapter);

        const doc: CdmDocumentDefinition = await cdmCorpus.fetchObjectAsync<CdmDocumentDefinition>('local:/missingNestedImport.cdm.json');
        expect(doc)
            .not
            .toBeUndefined();
        expect(doc.imports.length)
            .toBe(1);
        const firstImport: CdmDocumentDefinition = (doc.imports.allItems[0]).doc;
        expect(firstImport.imports.length)
            .toBe(1);
        expect(firstImport.name)
            .toBe('notMissing.cdm.json');
        const nestedImport: CdmDocumentDefinition = (firstImport.imports.allItems[0]).doc;
        expect(nestedImport)
            .toBeUndefined();
    });

    /**
     * Testing loading where import is listed multiple times in different files
     */
    it('TestEntityWithSameImportsAsync', async () => {
        const localAdapter: LocalAdapter = createStorageAdapterForTest('TestEntityWithSameImportsAsync');
        const cdmCorpus: CdmCorpusDefinition = createTestCorpus(localAdapter);

        const doc: CdmDocumentDefinition = await cdmCorpus.fetchObjectAsync<CdmDocumentDefinition>('local:/multipleImports.cdm.json');
        expect(doc)
            .not
            .toBeUndefined();
        expect(doc.imports.length)
            .toBe(2);
        const firstImport: CdmDocumentDefinition = (doc.imports.allItems[0]).doc;
        expect(firstImport.name)
            .toBe('missingImport.cdm.json');
        expect(firstImport.imports.length)
            .toBe(1);
        const secondImport: CdmDocumentDefinition = (doc.imports.allItems[1]).doc;
        expect(secondImport.name)
            .toBe('notMissing.cdm.json');
    });

    /**
     * Testing an import with a non-existing namespace name.
     */
    it('TestNonExistingAdapterNamespace', async () => {
        const localAdapter: LocalAdapter = createStorageAdapterForTest('TestNonExistingAdapterNamespace');
        const cdmCorpus: CdmCorpusDefinition = createTestCorpus(localAdapter);
        cdmCorpus.storage.mount('erp', localAdapter);

        // Set local as our default.
        cdmCorpus.storage.defaultNamespace = 'erp';

        const manifestPath: string = cdmCorpus.storage.createAbsoluteCorpusPath('erp.missingImportManifest.cdm');
        const rootManifest: CdmManifestDefinition = await cdmCorpus.createRootManifest(manifestPath);

        // Load a manifest that is trying to import from 'cdm' namespace.
        // The manifest does't exist since the import couldn't get resolved,
        // so the error message will be logged and the null value will be propagated back to a user.
        expect(rootManifest)
            .toBeUndefined();
    });

    /**
     * Testing docs that load the same import
     */
    it('TestLoadingSameImportsAsync', async () => {
        const localAdapter: LocalAdapter = createStorageAdapterForTest('TestLoadingSameImportsAsync');
        const cdmCorpus: CdmCorpusDefinition = createTestCorpus(localAdapter);

        const mainDoc: CdmDocumentDefinition = await cdmCorpus.fetchObjectAsync<CdmDocumentDefinition>('mainEntity.cdm.json');
        expect(mainDoc)
            .not
            .toBeUndefined();
        expect(mainDoc.imports.length)
            .toBe(2);
        const firstImport: CdmDocumentDefinition = (mainDoc.imports.allItems[0]).doc;
        const secondImport: CdmDocumentDefinition = (mainDoc.imports.allItems[1]).doc;

        // since these two imports are loaded asyncronously, we need to make sure that
        // the import that they share (targetImport) was loaded, and that the
        // targetImport doc is attached to both of these import objects
        expect(firstImport.imports.length)
            .toBe(1);
        expect(firstImport.imports.allItems[0].doc)
            .toBeDefined();

        expect(secondImport.imports.length)
            .toBe(1);
        expect(secondImport.imports.allItems[0].doc)
            .toBeDefined();
    });

    /**
     * Testing docs that load the same import of which, the file cannot be found
     */
    it('TestLoadingSameMissingImportsAsync', async () => {
        const localAdapter: LocalAdapter = createStorageAdapterForTest('TestLoadingSameMissingImportsAsync');
        const cdmCorpus: CdmCorpusDefinition = createTestCorpus(localAdapter);

        const mainDoc: CdmDocumentDefinition = await cdmCorpus.fetchObjectAsync<CdmDocumentDefinition>('mainEntity.cdm.json');
        expect(mainDoc)
            .not
            .toBeUndefined();
        expect(mainDoc.imports.length)
            .toBe(2);

        // make sure imports loaded correctly, despite them missing imports
        const firstImport: CdmDocumentDefinition = (mainDoc.imports.allItems[0]).doc;
        const secondImport: CdmDocumentDefinition = (mainDoc.imports.allItems[1]).doc;

        expect(firstImport.imports.length)
            .toBe(1);
        expect(firstImport.imports.allItems[0].doc)
            .toBeUndefined();

        expect(secondImport.imports.length)
            .toBe(1);
        expect(firstImport.imports.allItems[0].doc)
            .toBeUndefined();
    });

    /**
     * Testing doc that loads an import that has already been loaded before
     */
    it('TestLoadingAlreadyPresentImportsAsync', async () => {
        const localAdapter: LocalAdapter = createStorageAdapterForTest('TestLoadingAlreadyPresentImportsAsync');
        const cdmCorpus: CdmCorpusDefinition = createTestCorpus(localAdapter);

        // load the first doc
        const mainDoc: CdmDocumentDefinition = await cdmCorpus.fetchObjectAsync<CdmDocumentDefinition>('mainEntity.cdm.json');
        expect(mainDoc)
            .not
            .toBeUndefined();
        expect(mainDoc.imports.length)
            .toBe(1);

        const importDoc: CdmDocumentDefinition = (mainDoc.imports.allItems[0]).doc;
        expect(importDoc)
            .toBeDefined();

        // now load the second doc, which uses the same import
        // the import should not be loaded again, it should be the same object
        const secondDoc: CdmDocumentDefinition = await cdmCorpus.fetchObjectAsync<CdmDocumentDefinition>('secondEntity.cdm.json');
        expect(secondDoc)
            .not
            .toBeUndefined();
        expect(secondDoc.imports.length)
            .toBe(1);

        const secondImportDoc: CdmDocumentDefinition = (mainDoc.imports.allItems[0]).doc;
        expect(secondImportDoc)
            .toBeDefined();

        expect(importDoc)
            .toBe(secondImportDoc);
    });

    function createTestCorpus(adapter: StorageAdapter): CdmCorpusDefinition {
        const cdmCorpus: CdmCorpusDefinition = new CdmCorpusDefinition();
        cdmCorpus.storage.mount('local', adapter);
        cdmCorpus.storage.defaultNamespace = 'local';

        // Set empty callback to avoid breaking tests due too many errors in logs,
        // change the event callback to console or file status report if wanted.
        // tslint:disable-next-line: no-empty
        cdmCorpus.setEventCallback(() => { }, cdmStatusLevel.error);

        return cdmCorpus;
    }

    function createStorageAdapterForTest(testName: string): LocalAdapter {
        return new LocalAdapter(testHelper.getInputFolderPath(testsSubpath, testName));
    }
});

File: objectModel/Java/objectmodel/src/test/java/com/microsoft/commondatamodel/objectmodel/cdm/ImportsTest.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.cdm;

import static org.testng.Assert.assertNotNull;

import com.microsoft.commondatamodel.objectmodel.TestHelper;
import com.microsoft.commondatamodel.objectmodel.storage.LocalAdapter;
import com.microsoft.commondatamodel.objectmodel.storage.StorageAdapter;
import java.io.File;
import org.testng.Assert;
import org.testng.annotations.Test;

public class ImportsTest {
  /**
   * The path between TestDataPath and TestName.
   */
  private static final String TESTS_SUBPATH = new File("cdm", "imports").toString();

  @Test
  public void testEntityWithMissingImport() throws InterruptedException {
    final StorageAdapter localAdapter =
            this.createStorageAdapterForTest("testEntityWithMissingImport");

    final CdmCorpusDefinition cdmCorpus = this.createTestCorpus(localAdapter);

    final CdmDocumentDefinition doc =
            cdmCorpus.<CdmDocumentDefinition>fetchObjectAsync(
                    "local:/missingImport.cdm.json"
            ).join();
    assertNotNull(doc);
    Assert.assertEquals(doc.getImports().getCount(), 1);
    Assert.assertEquals(
            doc.getImports().get(0).getCorpusPath(),
            "missing.cdm.json");
    Assert.assertNull((doc.getImports().get(0)).getDoc());
  }

  @Test
  public void testEntityWithMissingNestedImports() throws InterruptedException {
    final StorageAdapter localAdapter =
            this.createStorageAdapterForTest("testEntityWithMissingNestedImports");

    final CdmCorpusDefinition cdmCorpus = this.createTestCorpus(localAdapter);

    final CdmDocumentDefinition doc =
            cdmCorpus.<CdmDocumentDefinition>fetchObjectAsync(
                    "local:/missingNestedImport.cdm.json"
            ).join();

    assertNotNull(doc);
    Assert.assertEquals(doc.getImports().getCount(), 1);
    final CdmDocumentDefinition firstImport = doc.getImports().getAllItems().get(0).getDoc();
    Assert.assertEquals(firstImport.getImports().getCount(), 1);
    Assert.assertEquals(firstImport.getName(), "notMissing.cdm.json");
    final CdmDocumentDefinition nestedImport = firstImport.getImports().getAllItems().get(0).getDoc();
    Assert.assertNull(nestedImport);
  }

  @Test
  public void testEntityWithSameImports() throws InterruptedException {
    final StorageAdapter localAdapter = this.createStorageAdapterForTest("testEntityWithSameImports");

    final CdmCorpusDefinition cdmCorpus = this.createTestCorpus(localAdapter);

    final CdmDocumentDefinition doc = cdmCorpus.<CdmDocumentDefinition>fetchObjectAsync("local:/multipleImports.cdm.json").join();

    assertNotNull(doc);
    Assert.assertEquals(doc.getImports().getCount(), 2);
    final CdmDocumentDefinition firstImport = doc.getImports().getAllItems().get(0).getDoc();
    Assert.assertEquals(firstImport.getName(), "missingImport.cdm.json");
    Assert.assertEquals(firstImport.getImports().getCount(), 1);
    final CdmDocumentDefinition secondImport = doc.getImports().getAllItems().get(1).getDoc();
    Assert.assertEquals(secondImport.getName(), "notMissing.cdm.json");
  }

  /**
   * Test an import with a non-existing namespace name.
   */
  @Test
  public void testNonExistingAdapterNamespace() throws InterruptedException {
    final StorageAdapter localAdapter = this.createStorageAdapterForTest("testNonExistingAdapterNamespace");

    final CdmCorpusDefinition cdmCorpus = this.createTestCorpus(localAdapter);

    // Register it as a 'local' adapter.
    cdmCorpus.getStorage().mount("erp", localAdapter);

    // Set local as our default.
    cdmCorpus.getStorage().setDefaultNamespace("erp");

    // Load a manifest that is trying to import from 'cdm' namespace.
    // The manifest does't exist since the import couldn't get resolved,
    // so the error message will be logged and the null value will be propagated back to a user.
    Assert.assertNull(cdmCorpus.<CdmManifestDefinition>fetchObjectAsync("erp.missingImportManifest.cdm").join());
  }

  /**
   * Testing docs that load the same import.
   */
  @Test
  public void testLoadingSameImports() throws InterruptedException {
    final StorageAdapter localAdapter = this.createStorageAdapterForTest("testLoadingSameImportsAsync");
    final CdmCorpusDefinition cdmCorpus = this.createTestCorpus(localAdapter);

    CdmDocumentDefinition mainDoc = cdmCorpus.<CdmDocumentDefinition>fetchObjectAsync("mainEntity.cdm.json").join();
    Assert.assertNotNull(mainDoc);
    Assert.assertEquals(mainDoc.getImports().getCount(), 2);

    CdmDocumentDefinition firstImport = mainDoc.getImports().get(0).getDoc();
    CdmDocumentDefinition secondImport = mainDoc.getImports().get(1).getDoc();

    // Since these two imports are loaded asynchronously, we need to make sure that
    // the import that they share (targetImport) was loaded, and that the
    // targetImport doc is attached to both of these import objects.
    Assert.assertEquals(firstImport.getImports().getCount(), 1);
    Assert.assertNotNull(firstImport.getImports().get(0).getDoc());
    Assert.assertEquals(secondImport.getImports().getCount(), 1);
    Assert.assertNotNull(secondImport.getImports().get(0).getDoc());
  }

  /**
   * Testing docs that load the same import.
   */
  @Test
  public void testLoadingSameMissingImports() throws InterruptedException {
    final StorageAdapter localAdapter = this.createStorageAdapterForTest("testLoadingSameMissingImportsAsync");
    final CdmCorpusDefinition cdmCorpus = this.createTestCorpus(localAdapter);

    CdmDocumentDefinition mainDoc = cdmCorpus.<CdmDocumentDefinition>fetchObjectAsync("mainEntity.cdm.json").join();
    Assert.assertNotNull(mainDoc);
    Assert.assertEquals(mainDoc.getImports().getCount(), 2);

    // Make sure imports loaded correctly, despite them missing imports.
    CdmDocumentDefinition firstImport = mainDoc.getImports().get(0).getDoc();
    CdmDocumentDefinition secondImport = mainDoc.getImports().get(0).getDoc();

    Assert.assertEquals(firstImport.getImports().getCount(), 1);
    Assert.assertNull(firstImport.getImports().get(0).getDoc());

    Assert.assertEquals(secondImport.getImports().getCount(), 1);
    Assert.assertNull(firstImport.getImports().get(0).getDoc());
  }

  /**
   * Testing docs that load the same import.
   */
  @Test
  public void testLoadingAlreadyPresentImports() throws InterruptedException {
    final StorageAdapter localAdapter = this.createStorageAdapterForTest("testLoadingAlreadyPresentImportsAsync");
    final CdmCorpusDefinition cdmCorpus = this.createTestCorpus(localAdapter);

    // Load the first doc.
    CdmDocumentDefinition mainDoc = cdmCorpus.<CdmDocumentDefinition>fetchObjectAsync("mainEntity.cdm.json").join();
    Assert.assertNotNull(mainDoc);
    Assert.assertEquals(mainDoc.getImports().getCount(), 1);

    CdmDocumentDefinition importDoc = mainDoc.getImports().get(0).getDoc();
    Assert.assertNotNull(importDoc);

    // Now load the second doc, which uses the same import.
    // The import should not be loaded again, it should be the same object.
    CdmDocumentDefinition secondDoc = cdmCorpus.<CdmDocumentDefinition>fetchObjectAsync("secondEntity.cdm.json").join();
    Assert.assertNotNull(secondDoc);
    Assert.assertEquals(secondDoc.getImports().getCount(), 1);

    CdmDocumentDefinition secondImportDoc = mainDoc.getImports().get(0).getDoc();
    Assert.assertNotNull(secondImportDoc);

    Assert.assertEquals(importDoc, secondImportDoc);
  }

  private CdmCorpusDefinition createTestCorpus(final StorageAdapter adapter) {
    final CdmCorpusDefinition cdmCorpus = new CdmCorpusDefinition();
    cdmCorpus.getStorage().mount("local", adapter);
    cdmCorpus.getStorage().setDefaultNamespace("local");

    return cdmCorpus;
  }

  /**
   * Creates a storage adapter used to retrieve input files associated with test.
   *
   * @param testName The name of the test we should retrieve input files for.
   * @return The storage adapter to be used by the named test method.
   */
  private StorageAdapter createStorageAdapterForTest(String testName) throws InterruptedException {
    return new LocalAdapter(TestHelper.getInputFolderPath(TESTS_SUBPATH, testName));
  }
}
